src/python/pants/backend/javascript/nodejs_project.py

```diff
@@ -21,26 +29,129 @@
 
 
 @dataclass(frozen=True)
-class NodeJSProject:
+class _TentativeProject:
     root_dir: str
     workspaces: FrozenOrderedSet[PackageJson]
     default_resolve_name: str
 
-    def is_parent(self, project: NodeJSProject) -> bool:
+    def is_parent(self, project: _TentativeProject) -> bool:
         return self.root_dir != project.root_dir and any(
             project.root_dir == workspace.root_dir for workspace in self.workspaces
         )
 
-    def including_workspaces_from(self, child: NodeJSProject) -> NodeJSProject:
+    def including_workspaces_from(self, child: _TentativeProject) -> _TentativeProject:
         return replace(self, workspaces=self.workspaces | child.workspaces)
 
+    def root_workspace(self) -> PackageJson | None:
+        for ws in self.workspaces:
+            if ws.root_dir == self.root_dir:
+                return ws
+        return None
+
+
+@dataclass(frozen=True)
+class NodeJSProject:
```

Comment:
Should this be implemented via a union?

I'm hesitant because I don't think users will want to extend this union themselves, but it might make the code cleaner w.r.t to the package manager specific properties.

Reply:
While it's totally fine to have "private" `@unions`, I don't think that seems necessary yet, no. And it's probably worth avoiding prematurely extracting interfaces in case #16763 comes along and makes it much easier to use `@union`s by allowing methods to be awaited.

But when in doubt, leaving a TODO comment behind posing the question can be helpful, as it might help the reader come to another conclusion.